# Data Overview and Proposed Work Plan

In [19]:
import pandas as pd

In [20]:
# Create file paths
contract_path = "/Users/sabrinamcfield/Data_Science_Projects-TripleTen-/project2_churn_prediction/final_provider/contract.csv"
personal_path = "/Users/sabrinamcfield/Data_Science_Projects-TripleTen-/project2_churn_prediction/final_provider/personal.csv"
internet_path = "/Users/sabrinamcfield/Data_Science_Projects-TripleTen-/project2_churn_prediction/final_provider/internet.csv"
phone_path = "/Users/sabrinamcfield/Data_Science_Projects-TripleTen-/project2_churn_prediction/final_provider/phone.csv"

# Load datasets
contract_df = pd.read_csv(contract_path)
personal_df = pd.read_csv(personal_path)
internet_df = pd.read_csv(internet_path)
phone_df = pd.read_csv(phone_path)

In [7]:
# Get a general overview of the data
contract_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   BeginDate         7043 non-null   object 
 2   EndDate           7043 non-null   object 
 3   Type              7043 non-null   object 
 4   PaperlessBilling  7043 non-null   object 
 5   PaymentMethod     7043 non-null   object 
 6   MonthlyCharges    7043 non-null   float64
 7   TotalCharges      7043 non-null   object 
dtypes: float64(1), object(7)
memory usage: 440.3+ KB


In [8]:
# View the data
contract_df

customerID   BeginDate              EndDate            Type  \
0     7590-VHVEG  2020-01-01                   No  Month-to-month   
1     5575-GNVDE  2017-04-01                   No        One year   
2     3668-QPYBK  2019-10-01  2019-12-01 00:00:00  Month-to-month   
3     7795-CFOCW  2016-05-01                   No        One year   
4     9237-HQITU  2019-09-01  2019-11-01 00:00:00  Month-to-month   
...          ...         ...                  ...             ...   
7038  6840-RESVB  2018-02-01                   No        One year   
7039  2234-XADUH  2014-02-01                   No        One year   
7040  4801-JZAZL  2019-03-01                   No  Month-to-month   
7041  8361-LTMKD  2019-07-01  2019-11-01 00:00:00  Month-to-month   
7042  3186-AJIEK  2014-08-01                   No        Two year   

     PaperlessBilling              PaymentMethod  MonthlyCharges TotalCharges  
0                 Yes           Electronic check           29.85        29.85  
1                  No               Mailed check           56.95       1889.5  
2                 Yes               Mailed check           53.85       108.15  
3                  No  Bank transfer (automatic)           42.30      1840.75  
4                 Yes           Electronic check           70.70       151.65  
...               ...                        ...             ...          ...  
7038              Yes               Mailed check           84.80       1990.5  
7039              Yes    Credit card (automatic)          103.20       7362.9  
7040              Yes           Electronic check           29.60       346.45  
7041              Yes               Mailed check           74.40        306.6  
7042              Yes  Bank transfer (automatic)          105.65       6844.5  

[7043 rows x 8 columns]

In [9]:
# Find the percentage of current customers
len(contract_df[contract_df['EndDate'] == 'No'])/ len(contract_df['EndDate'])

0.7346301292063041

#### **Notes About Contract Dataset**
In the contract dataset, we have 7043 customers accounted for. About 73% of customers have not turned which indicates we have an imbalanced class. We also have no missing data.

**Preprocessing Tasks**
- Column names and string values need to be standardized to a lowercase format, remove whitespace and use of underscore to separate words.
- Change 'TotalCharges' column into a float64
- Convert 'BeginDate' to a datetime format

**Feature Engineering Tasks**
- 'Type', 'PaperlessBilling' and 'PaymentMethod' will need to be converted categorically using label encoding (paperless billing), one hot encoding (payment methode) and ordinal encoding (type)
- 'EndDate' is a binary target and should be defined as 1 = churn and 0 = no_churn in a new column called 'churn'
- Check seasonalty for BeginDate to identify if they only signed up for the deal and add 'contract_age' column
- Perform class balancing on the training set after the split and before model training

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> Nice job writing out your to-dos.
<a class="tocSkip"></a>

In [10]:
personal_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customerID     7043 non-null   object
 1   gender         7043 non-null   object
 2   SeniorCitizen  7043 non-null   int64 
 3   Partner        7043 non-null   object
 4   Dependents     7043 non-null   object
dtypes: int64(1), object(4)
memory usage: 275.2+ KB


In [11]:
personal_df

customerID  gender  SeniorCitizen Partner Dependents
0     7590-VHVEG  Female              0     Yes         No
1     5575-GNVDE    Male              0      No         No
2     3668-QPYBK    Male              0      No         No
3     7795-CFOCW    Male              0      No         No
4     9237-HQITU  Female              0      No         No
...          ...     ...            ...     ...        ...
7038  6840-RESVB    Male              0     Yes        Yes
7039  2234-XADUH  Female              0     Yes        Yes
7040  4801-JZAZL  Female              0     Yes        Yes
7041  8361-LTMKD    Male              1     Yes         No
7042  3186-AJIEK    Male              0      No         No

[7043 rows x 5 columns]

#### **Notes About Personal Dataset**
In the personal dataset, we have 7043 customers accounted for. We also have no missing data.

**Preprocessing Tasks**
- Column names need to be standardized to a lowercase format, remove whitespace and use of underscore to separate words.
- String values need to be standardized to a lowercase format


**Feature Engineering Tasks**
- 'gender', 'Partner' and 'Dependents' will need to be converted categorically using label encoding
- After preprocessing, merge with contract_df on 'customerID'

In [12]:
internet_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5517 entries, 0 to 5516
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   customerID        5517 non-null   object
 1   InternetService   5517 non-null   object
 2   OnlineSecurity    5517 non-null   object
 3   OnlineBackup      5517 non-null   object
 4   DeviceProtection  5517 non-null   object
 5   TechSupport       5517 non-null   object
 6   StreamingTV       5517 non-null   object
 7   StreamingMovies   5517 non-null   object
dtypes: object(8)
memory usage: 344.9+ KB


In [13]:
internet_df

customerID InternetService OnlineSecurity OnlineBackup DeviceProtection  \
0     7590-VHVEG             DSL             No          Yes               No   
1     5575-GNVDE             DSL            Yes           No              Yes   
2     3668-QPYBK             DSL            Yes          Yes               No   
3     7795-CFOCW             DSL            Yes           No              Yes   
4     9237-HQITU     Fiber optic             No           No               No   
...          ...             ...            ...          ...              ...   
5512  6840-RESVB             DSL            Yes           No              Yes   
5513  2234-XADUH     Fiber optic             No          Yes              Yes   
5514  4801-JZAZL             DSL            Yes           No               No   
5515  8361-LTMKD     Fiber optic             No           No               No   
5516  3186-AJIEK     Fiber optic            Yes           No              Yes   

     TechSupport StreamingTV StreamingMovies  
0             No          No              No  
1             No          No              No  
2             No          No              No  
3            Yes          No              No  
4             No          No              No  
...          ...         ...             ...  
5512         Yes         Yes             Yes  
5513          No         Yes             Yes  
5514          No          No              No  
5515          No          No              No  
5516         Yes         Yes             Yes  

[5517 rows x 8 columns]

In [14]:
internet_df['customerID'].value_counts()

customerID
7590-VHVEG    1
5181-OABFK    1
9356-AXGMP    1
5176-OLSKT    1
7508-DQAKK    1
             ..
1104-FEJAM    1
2972-YDYUW    1
4238-JSSWH    1
0869-PAPRP    1
3186-AJIEK    1
Name: count, Length: 5517, dtype: int64

#### **Notes About Internet Dataset**
In the internet dataset, we have 5517 out of 7043 customers accounted for using internet. We also have no missing data.  

**Preprocessing Tasks**
- Standardize the column names 

**Feature Engineering Tasks**
- Create binary values for columns with yes/no values so that yes = 1 and no = 0

**Clarifying Questions**
- Should I create an additional binary column (e.g., has_internet) to indicate whether the customer has any internet service?

In [15]:
phone_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6361 entries, 0 to 6360
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customerID     6361 non-null   object
 1   MultipleLines  6361 non-null   object
dtypes: object(2)
memory usage: 99.5+ KB


In [16]:
phone_df

customerID MultipleLines
0     5575-GNVDE            No
1     3668-QPYBK            No
2     9237-HQITU            No
3     9305-CDSKC           Yes
4     1452-KIOVK           Yes
...          ...           ...
6356  2569-WGERO            No
6357  6840-RESVB           Yes
6358  2234-XADUH           Yes
6359  8361-LTMKD           Yes
6360  3186-AJIEK            No

[6361 rows x 2 columns]

In [17]:
# Check for duplicates
phone_df['customerID'].duplicated().sum()

0

In [18]:
# Find customerIDs in contract_df that are not in phone_df
missing_phone_customers = contract_df[~contract_df['customerID'].isin(phone_df['customerID'])]

# Display only the customerIDs
missing_ids = missing_phone_customers['customerID']
print(missing_ids)

0       7590-VHVEG
3       7795-CFOCW
7       6713-OKOMC
20      8779-QRDMV
27      8665-UTDHZ
           ...    
7003    4501-VCPFK
7007    2274-XUATA
7029    2235-DWLJU
7036    7750-EYXWZ
7040    4801-JZAZL
Name: customerID, Length: 682, dtype: object


#### **Notes About Internet Dataset**
In the phone dataset, we have 6361 out of 7043 customers accounted for having multiple phone lines or not. There are no columns to distinguish if the customer has phone service or not. We also have no missing data. Immediately I noticed there are no columns titled 'customerID' like the other datasets. 

**Preprocessing Tasks**
- Column names need to be standardized to lowercase and add underscores

**Feature Engineering Tasks**
- the 'MultipleLines' column needs a categorical encoding like the others. I will use one-hot encoding for this column.

**Clarifying Questions**
- What does 'MultipleLines' actually represent? Does it refer to multiple phone lines under one account, or multiple users on one line?

- How should the 682 customers without multiple lines be treated — included with missing values or excluded from the analysis?

# **Proposed Work Plan**
The goal of this project is to develop a model that will determine if a customer will leave. These are the steps that I will take to accomplish this task;

1. Download the data
2. Explore the data to determine how to handle it in preprocessing
3. Perform preprocessing on the data to include;
    - Merge all datasets
    - Standardize the format to lowercase with underscores on columns and values
    - Convert datatypes to most appropriate type for feature engineering
      
5. Perform feature engineering;
     - Before merging, create a user_id column for all datasets based off of the first 4 characters in the customerID then merge on user_id.
     - Create new features:
         - 'Churn' with 1 = churn and 0 = not churn
         - 'Tenure' from 'BeginDate'
         - 'Long_Term_Contract' = contract != 'month-to-month'
     - Encode categorical variables:
         - Use Label encoding for binary features: 'gender', 'Partner', 'Dependents', 'PaperlessBilling'
         - Ordinal encoding for ordered categories: 'Type'
         - One-hot encoding: 'PaymentMethod', 'MultipleLines', 'InternetService'

8. Fill in missing values that occurred due to the merge with 'unknown'
9. Perform EDA (Exploratory Data Analysis)
    - Visualize churn distribution
    - Analyze relationships between features and churn
    - Identify imbalance and correlation
11. Model training and evaluation
    - Split data: training, validation, and test sets (e.g., 70/15/15).
    - Handle class imbalance:
        - Use class_weight='balanced' or apply SMOTE on the training set
    - Train and compare multiple models:
        - Logistic Regression (baseline)
        - Random Forest
        - XGBoost or LightGBM (for performance)
    - Tune hyperparameters using GridSearchCV or RandomizedSearchCV
    - Evaluate with:
        - AUC-ROC (primary)
        - Accuracy, precision, recall (secondary)
12. Write a conclusion